### PR TITLE
Tweak spans for trait bounds on associated types

### PR DESCRIPTION
**Pull request by estebank**
Provide more information on projection failure, by pointing at the associated type (the note on line 16 and the use of a note instead of a label on line 27 are new):

```
error[E0271]: type mismatch resolving `<ModelT as Vehicle>::Color == Blue`
  --> src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.rs:31:10
   |
31 | fn b() { blue_car(ModelT); } //~ ERROR type mismatch
   |          ^^^^^^^^ type mismatch resolving `<ModelT as Vehicle>::Color == Blue`
   |
note: expected this to be `Blue`
  --> src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.rs:16:40
   |
16 | impl Vehicle for ModelT { type Color = Black; }
   |                                        ^^^^^
note: required by a bound in `blue_car`
  --> src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.rs:27:19
   |
27 | fn blue_car<C:Car<Color=Blue>>(c: C) {
   |                   ^^^^^^^^^^ required by this bound in `blue_car`
```

We are now pointing at the two places where the user can change things to make them compile: the trait bound and the associated type. We can't know for sure which of the two would be more appropriate on each case (although we already provide some suggestions when an extra trait bound would fix things), so instead we just put all the cards on the table and let the user decide for themselves with all the information at their disposal.

Reminded about this by the recent closing of #57663.

### CODE REVIEW

File: src/test/ui/associated-types/issue-63593.stderr [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
+  --> $DIR/issue-63593.rs:9:17
    |
 LL |     type This = Self;
-   |     ^^^^^^^^^^^^^^^^^
-   |     |
-   |     doesn't have a size known at compile-time
+   |     ------------^^^^-
+   |     |           |
+   |     |           doesn't have a size known at compile-time
    |     required by this bound in `MyTrait::This`
```
Comment (estebank) [inline src/test/ui/associated-types/issue-63593.stderr:8]: This one would be interesting to improve by adding a new Obligation Cause Code specific to the implicit `Sized` obligation for associated types and explain that with a note.

File: src/test/ui/generic-associated-types/issue-47206-where-clause.stderr [hunk 2]
```diff
@@ -1,11 +1,11 @@
 error[E0276]: impl has stricter requirements than trait
-  --> $DIR/issue-47206-where-clause.rs:13:5
+  --> $DIR/issue-47206-where-clause.rs:13:40
    |
 LL |     type Assoc3<T>;
    |     --------------- definition of `Assoc3` from trait
 ...
 LL |     type Assoc3<T> where T: Iterator = Vec<T>;
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl has extra requirement `T: Iterator`
+   |                                        ^^^^^^ impl has extra requirement `T: Iterator`
```
Comment (estebank) [inline src/test/ui/generic-associated-types/issue-47206-where-clause.stderr:8]: This has an incorrect span, this should point at either the whole assoc type _or_, even better yet, at the `T: Iterator` extra requirement. For that last part we'd need specific logic doing that.

File: src/test/ui/generic-associated-types/issue-68656-unsized-values.stderr [hunk 3]
```diff
@@ -8,15 +8,15 @@ LL | #![feature(generic_associated_types)]
...
+  --> $DIR/issue-68656-unsized-values.rs:16:21
    |
 LL |     type Item<'a>: std::ops::Deref<Target = T>;
    |                                    ---------- required by this bound in `UnsafeCopy::Item`
 ...
 LL | impl<T: Copy + std::ops::Deref> UnsafeCopy<T> for T {
    |      - this type parameter
 LL |     type Item<'a> = T;
-   |     ^^^^^^^^^^^^^^^^^^ expected type parameter `T`, found associated type
+   |                     ^ expected type parameter `T`, found associated type
```
Comment (estebank) [inline src/test/ui/generic-associated-types/issue-68656-unsized-values.stderr:19]: This seems somewhat confusing, but thankfully we provide a structured suggestion.

File: src/test/ui/hrtb/issue-62203-hrtb-ice.stderr [hunk 4]
```diff
@@ -1,8 +1,11 @@
 error[E0271]: type mismatch resolving `for<'r> <L<[closure@$DIR/issue-62203-hrtb-ice.rs:42:17: 42:39]> as T0<'r, (<Unit2 as Ty<'r>>::V,)>>::O == <_ as Ty<'r>>::V`
   --> $DIR/issue-62203-hrtb-ice.rs:38:19
    |
+LL |     type O = T::Output;
+   |              --------- expected this to be `Unit4`
+...
 LL |     let v = Unit2.m(
-   |                   ^ expected struct `Unit4`, found associated type
+   |                   ^ type mismatch resolving `for<'r> <L<[closure@$DIR/issue-62203-hrtb-ice.rs:42:17: 42:39]> as T0<'r, (<Unit2 as Ty<'r>>::V,)>>::O == <_ as Ty<'r>>::V`
```
Comment (estebank) [inline src/test/ui/hrtb/issue-62203-hrtb-ice.stderr:8]: That's quite a mouthful 😬

**Comment by hellow554** [comment]
You are truly amazing. Thank you for making the suggestion even better! ❤️

File: src/test/ui/associated-types/issue-63593.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/associated-types/issue-63593.stderr:8]: This is being addressed in https://github.com/rust-lang/rust/pull/85947

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 5]
```diff
@@ -1428,6 +1428,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         secondary_span: Option<(Span, String)>,
         mut values: Option<ValuePairs<'tcx>>,
         terr: &TypeError<'tcx>,
+        swap_secondary_and_primary: bool,
```
Comment (pnkfelix) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:1431]: hmm, should this have a comment with some guidance about when the `swap_secondary_and_primary` argument should be `true`? Or I guess we can rely on people just trying it out and seeing how it looks?

Comment (estebank) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs:1431]: Sure, I can add a docstring mentioning the function, but I would expect that "see how it affects the output" would be _my_ natural inclination.

File: src/test/ui/associated-types/issue-63593.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/associated-types/issue-63593.stderr]: Addressed

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs:1440]: Added a docstring to the method with some info.

**Comment by varkor** [comment]
I don't have time to review this right now. Going to reassign to r? @pnkfelix, as they've already taken a look at it. Apologies for taking so long to get around to this.

File: src/test/ui/issues/issue-73886.stderr [hunk 6]
```diff
@@ -4,14 +4,19 @@ error[E0605]: non-primitive cast: `&&[i32; 1]` as `&[_]`
...
 LL |     let _ = 7u32 as Option<_>;
-   |             ^^^^^^^^^^^^^^^^^ help: consider using the `From` trait instead: `Option<_>::from(7u32)`
+   |             ^^^^^^^^^^^^^^^^^
    |
-   = note: an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
+help: consider using a box or reference as appropriate
+  --> $DIR/issue-73886.rs:4:13
+   |
+LL |     let _ = 7u32 as Option<_>;
+   |             ^^^^
```
Comment (estebank) [inline src/test/ui/issues/issue-73886.stderr:17]: This is a regression, but seeing the way _other_ error improve, I wouldn't want to block on it.

File: src/test/ui/range/range-1.stderr [hunk 7]
```diff
@@ -19,15 +19,26 @@ error[E0277]: the size for values of type `[{integer}]` cannot be known at compi
...
+  --> $DIR/range-1.rs:14:9
    |
-   = help: the trait `Sized` is not implemented for `[{integer}]`
+LL |     let range = *arr..;
+   |         ^^^^^ doesn't have a size known at compile-time
+   |
+   = help: within `RangeFrom<[{integer}]>`, the trait `Sized` is not implemented for `[{integer}]`
+   = note: required because it appears within the type `RangeFrom<[{integer}]>`
+   = note: all local variables must have a statically known size
+   = help: unsized locals are gated as an unstable feature
```
Comment (estebank) [inline src/test/ui/range/range-1.stderr:39]: Arguably a mild regression (unnecessary verbosity).

File: compiler/rustc_hir/src/hir.rs [hunk 8]
```diff
@@ -401,6 +401,7 @@ pub enum GenericBound<'hir> {
     Trait(PolyTraitRef<'hir>, TraitBoundModifier),
     // FIXME(davidtwco): Introduce `PolyTraitRef::LangItem`
     LangItemTrait(LangItem, Span, HirId, &'hir GenericArgs<'hir>),
+    Unsized(Span),
```
Comment (tlyu) [inline compiler/rustc_hir/src/hir.rs:404]: Hm, I guess leaving off the `HirId` completely is one way to work around breaking the HIR invariant. Will the lack of `HirId` here cause problems later on, though?

Comment (estebank) [reply compiler/rustc_hir/src/hir.rs:404]: I'm really not sure. It _seems_ to me that it shouldn't be a problem, but could envision a future refactor causing issues with this approach. This variant is only used for signaling at the moment (we used to just drop this case on the floor early, as seen on `compiler/rustc_ast_lowering/src/item.rs`).

File: src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.stderr [hunk 9]
```diff
@@ -1,20 +1,26 @@
 error[E0271]: type mismatch resolving `<ModelT as Vehicle>::Color == Blue`
   --> $DIR/associated-types-binding-to-type-defined-in-supertrait.rs:31:10
    |
+LL | impl Vehicle for ModelT { type Color = Black; }
+   |                                        ----- expected this to be `Blue`
```
Comment (Aaron1011) [inline src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.stderr:5]: Personally, I find error messages hard to follow when they start with 'context', as opposed to the location where the error occurs. I think a better order would be:

`blue_var(ModelT) // type mismatch resolving`
`fn blue_car // required by this bound`
`impl Vehicle for ModelT` // expected this to be `lue`

Comment (estebank) [reply src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.stderr:5]: We could change the order by leveraging a new `note` instead of going straight to `span_label`s...

Comment (estebank) [reply src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.stderr:5]: Is this closer to what you'd desire?
linked error using notes instead of labels

Comment (estebank) [reply src/test/ui/associated-types/associated-types-binding-to-type-defined-in-supertrait.stderr]: Added in the last commit, but I feel like we should separate this in a different PR, to make it easier for the reviewer.

**Comment by estebank** [comment]
It seems like the performance impact might be driven by an increase in `evaluate_obligation` from 1.85% to 2.79% and `type_op_prove_predicate` from 4.35% to 5.69%. I will see if I can bring the performance in check, but we *are* tracking more things now :-/

File: compiler/rustc_infer/src/infer/outlives/obligations.rs [hunk 10]
```diff
@@ -99,7 +99,15 @@ impl<'cx, 'tcx> InferCtxt<'cx, 'tcx> {
...
-            infer::RelateParamBound(cause.span, sup_type)
+            infer::RelateParamBound(
+                cause.span,
+                sup_type,
+                match cause.code.peel_derives() {
+                    ObligationCauseCode::BindingObligation(_, span)
+                    | ObligationCauseCode::ImplicitSizedObligation(_, span) => Some(*span),
+                    _ => None,
+                },
+            )
```
Comment (estebank) [inline compiler/rustc_infer/src/infer/outlives/obligations.rs:110]: I suspect that this might be one of the reasons for the seen performance hit.

Comment (oli-obk) [reply compiler/rustc_infer/src/infer/outlives/obligations.rs:110]: This seems fairly cheap to me, the span is only used in the error case after all. The rest of the variants are larger, so adding the additional span didn't cause a  size increase

File: compiler/rustc_ast_lowering/src/item.rs [hunk 11]
```diff
@@ -1465,16 +1465,19 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+                                // parameters already, but we need to keep them around only for
+                                // diagnostics when we suggest removal of `?Sized` bounds. See
+                                // `suggest_constraining_type_param`. This will need to change if
+                                // we ever allow something *other* than `?Sized`.
+                                GenericBound::Trait(p, TraitBoundModifier::Maybe) => {
+                                    hir::GenericBound::Unsized(p.span)
+                                }
+                                _ => this.lower_param_bound(bound, ImplTraitContext::disallowed()),
+                            },
+                        )),
```
Comment (oli-obk) [inline compiler/rustc_ast_lowering/src/item.rs:1480]: This could be the source of the perf regression. Can you do a PR with just that change and run it through perf? If it is the problem, we likely just need a few early aborts for `?Sized` bounds so we don't even call queries on it and just immediately return the right value.

File: compiler/rustc_middle/src/ty/mod.rs [hunk 12]
```diff
@@ -458,6 +458,11 @@ pub enum PredicateKind<'tcx> {
     /// `const fn foobar<Foo: Bar>() {}`).
     Trait(TraitPredicate<'tcx>, Constness),
 
+    /// Equivalent to `Trait(SizedTraitPredicate, Constness::NotConst)` that type parameters and
+    /// associated types implicitly have. Used to differentiate them when giving trait bound
+    /// errors.
+    ImplicitSizedTrait(TraitPredicate<'tcx>),
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/mod.rs:464]: maybe just add a field to `TraitPredicate`? The additional variant requires a lot of noisy changes throughout the compiler.

Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs:464]: Adding a field to `TraitPredicate` caused fingerprinting issues that went away once I added a new variant.

**Comment by estebank** [comment]
@rust-timer queue

File: compiler/rustc_middle/src/ty/mod.rs (hunk 12 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/ty/mod.rs:464]: that's odd, do we have any manual eq/hash/stablehash impls on TraitPredicate?

**Comment by jackh726** [comment]
I don't have time to review this right now, but I have some concerns about this approach here (specifically e.g. adding the new PredicateKind). I'll try to elaborate more this weekend.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 12 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs:464]: Now I remember what I did: I `impl Eq for TraitPredicate` so that `Sized` obligations would always eq regardless of provenance (which seeing that the only place that that is needed is in `compiler/rustc_trait_selection/src/traits/auto_trait.rs`, it might be fine to add a separate method for "agnostic equality". The problem I imagined was that in the future someone might make a change while not being aware of the issues that doesn't account for the existence of a separate method instead of just `==`.

**Comment by estebank** [comment]
Split some changes from here onto #91769. Pushed the previous state of this PR to https://github.com/rust-lang/rust/commit/8dbee5ebfdac4e09d894b90f52205397851a5e3c.

Removed `ImplicitSizedObligation` and added a field to track this like we do for constness.

File: src/test/ui/wf/wf-trait-superbound.stderr [hunk 13]
```diff
@@ -4,7 +4,7 @@ error[E0277]: the trait bound `T: Copy` is not satisfied
 LL | trait SomeTrait<T>: ExtraCopy<T> {
    |                     ^^^^^^^^^^^^ the trait `Copy` is not implemented for `T`
    |
-note: required by a bound in `ExtraCopy`
+note: type parameters have an implicit `Sized` obligation
```
Comment (estebank) [inline src/test/ui/wf/wf-trait-superbound.stderr:7]: It seems like some tweaking will be needed :-/

Comment (estebank) [reply src/test/ui/wf/wf-trait-superbound.stderr]: Fixed, wasn't properly filtering these in `wf`.

File: src/test/ui/traits/mutual-recursion-issue-75860.stderr [hunk 14]
```diff
@@ -5,7 +5,135 @@ LL |     iso(left, right)
    |     ^^^
    |
    = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`mutual_recursion_issue_75860`)
-note: required by a bound in `Option`
+   = note: required because it appears within the type `Option<Option<_>>`
```
Comment (estebank) [inline src/test/ui/traits/mutual-recursion-issue-75860.stderr:8]: This will need to be fixed before merging. Pretty sure it is part of the other PR.

File: src/test/ui/range/range-1.stderr [hunk 15]
```diff
@@ -20,13 +20,24 @@ LL |     let range = *arr..;
...
+error[E0277]: the size for values of type `[{integer}]` cannot be known at compilation time
+  --> $DIR/range-1.rs:14:9
+   |
+LL |     let range = *arr..;
+   |         ^^^^^ doesn't have a size known at compile-time
+   |
+   = help: within `RangeFrom<[{integer}]>`, the trait `Sized` is not implemented for `[{integer}]`
+   = note: required because it appears within the type `RangeFrom<[{integer}]>`
+   = note: all local variables must have a statically known size
+   = help: unsized locals are gated as an unstable feature
```
Comment (estebank) [inline src/test/ui/range/range-1.stderr:38]: Acceptable verbosity regression, IMO.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 16]
```diff
@@ -1960,9 +1963,65 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 let sp = tcx.sess.source_map().guess_head_span(sp);
                 err.span_note(sp, &msg);
             }
+            ObligationCauseCode::ImplicitSizedObligation(item_def_id, span) => {
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1965]: @oli-obk let me know given the new changes whether this still applies.

File: compiler/rustc_middle/src/ty/mod.rs [hunk 17]
```diff
@@ -165,6 +165,25 @@ pub struct ImplHeader<'tcx> {
     pub predicates: Vec<Predicate<'tcx>>,
 }
 
+#[derive(
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/mod.rs:168]: This would probably look better as two separate derive lines...

File: compiler/rustc_middle/src/ty/mod.rs [hunk 18]
```diff
@@ -165,6 +165,25 @@ pub struct ImplHeader<'tcx> {
...
+    PartialEq,
+    Eq,
+    Hash,
+    TyEncodable,
+    TyDecodable,
+    HashStable,
+    Debug,
+    TypeFoldable
+)]
+pub enum ImplicitBound {
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/mod.rs:180]: Why isn't this just a "Yes" or "No"? `Sized` should be inferred from the trait def id?

**Review by jackh726 — COMMENTED** [review]
I've only briefly glanced over this, but already looks like a better approach than previously.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 18 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs:180]: This could potentially be extended to support auto-traits in general, but it can easily be changed.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 17 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs:168]: I was just following what rustfmt/existing code was doing.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 18 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs:180]: Changed to `Yes`.

File: compiler/rustc_middle/src/ty/relate.rs [hunk 19]
```diff
@@ -788,6 +788,21 @@ impl<'tcx> Relate<'tcx> for GenericArg<'tcx> {
...
+        a: ty::ImplicitBound,
+        b: ty::ImplicitBound,
+    ) -> RelateResult<'tcx, ty::ImplicitBound> {
+        match (a, b) {
+            (ty::ImplicitBound::No, ty::ty::ImplicitBound::No) => Ok(ty::ImplicitBound::No),
+            (ty::ImplicitBound::Yes, ty::ty::ImplicitBound::Yes)
+            | (ty::ImplicitBound::Yes, ty::ty::ImplicitBound::No)
+            | (ty::ImplicitBound::No, ty::ty::ImplicitBound::Yes) => Ok(ty::ImplicitBound::Yes),
+        }
+    }
```
Comment (estebank) [inline compiler/rustc_middle/src/ty/relate.rs:803]: Please take a second look at this, I wouldn't be surprised if this is wrong.

**Comment by estebank** [comment]
I'm at a loss on how my changes can cause `type_op_prove_predicate` and `evaluate_obligation` are regressing *so much*, the code is very straightforward.

**Comment by jackh726** [comment]
Going to mark this as waiting on author. Seems like it needs a rebase and perf fix.

**Comment by Dylan-DPC** [comment]
Closing this as it has bitrotted
